# Step 4: Correlate Category Embeddings

This notebook correlates category-level average embeddings between two embedding files (e.g., bv_clip and things_clip), computes **CLIP vs DINOv3** agreement across categories, and generates BV–THINGS correlation montages and histograms.

## Overview

This step:
1. Loads category average embeddings from two sources
2. Finds matching categories between the two sets
3. Computes correlations (Pearson, Spearman, Cosine) for each category
4. Reports summary statistics and top/bottom categories
5. **CLIP vs DINOv3 correlation**: how correlated are category-wise BV–THINGS correlations between CLIP and DINOv3? (e.g., do categories with high CLIP agreement also have high DINOv3 agreement?)
6. **BV–THINGS montages**: for CLIP and DINOv3, generates figures with high/low correlation category montages and correlation histograms

## Prerequisites

This step requires:
- Output from Step 2 (e.g., bv_clip and things_clip category average embeddings)
- (Optional) image dirs for montages

## Setup and Imports

In [1]:
import re
import numpy as np
import pandas as pd
from pathlib import Path
from scipy.stats import pearsonr, spearmanr, kendalltau
from sklearn.metrics.pairwise import cosine_similarity
import matplotlib
matplotlib.use("Agg")
import matplotlib.pyplot as plt
from matplotlib.gridspec import GridSpec
from PIL import Image
import warnings
warnings.filterwarnings('ignore')

print("All imports successful!")

All imports successful!


## Configuration

**Please update the paths below according to your setup:**

In [2]:
# ============================================================================
# CONFIGURATION - UPDATE THESE PATHS FOR YOUR SETUP
# ============================================================================
NOTEBOOK_DIR = Path(".").resolve()  # run from analysis/vss-2026
DATA_ROOT = NOTEBOOK_DIR.parent.parent  # repo root (object-detection) for data/

# --- Category embeddings & correlation ---
COMP_DIR = NOTEBOOK_DIR / "bv_things_comp_12252025"
# One model pair for the main correlation loop; can run twice (CLIP, DINOv3) by changing these
CSV_FILE1 = COMP_DIR / "bv_dinov3_filtered_zscored_hierarchical_163cats/normalized_filtered_embeddings_alphabetical.csv"
CSV_FILE2 = COMP_DIR / "things_dinov3_filtered_zscored_hierarchical_163cats/normalized_filtered_embeddings_alphabetical.csv"
# All four for montages (CLIP + DINOv3)
BV_CLIP_CSV = COMP_DIR / "bv_clip_filtered_zscored_hierarchical_163cats/normalized_filtered_embeddings_alphabetical.csv"
THINGS_CLIP_CSV = COMP_DIR / "things_clip_filtered_zscored_hierarchical_163cats/normalized_filtered_embeddings_alphabetical.csv"
BV_DINOV3_CSV = COMP_DIR / "bv_dinov3_filtered_zscored_hierarchical_163cats/normalized_filtered_embeddings_alphabetical.csv"
THINGS_DINOV3_CSV = COMP_DIR / "things_dinov3_filtered_zscored_hierarchical_163cats/normalized_filtered_embeddings_alphabetical.csv"

CATEGORY_INCLUDE_FILE = DATA_ROOT / "data/things_bv_overlap_categories_exclude_zero_precisions.txt"
OUTPUT_DIR = NOTEBOOK_DIR / "bv_things_comp_12252025/correlation_results_12252025"
OUTPUT_FILENAME = 'bv_things_dinov3_category_embeddings_correlations.txt'
# Output for CLIP vs DINOv3 category correlation table
CLIP_VS_DINOV3_CSV = OUTPUT_DIR / "clip_vs_dinov3_category_correlations.csv"

# --- BV–THINGS montages (optional; set to None or ensure dirs exist) ---
CROPPED_DIR = Path("/data2/dataset/babyview/868_hours/outputs/yoloe_cdi_all_cropped_by_class")
THINGS_IMAGES_DIR = Path("/ccn2/dataset/babyview/outputs_20250312/things_bv_overlapping_categories_corrected")
CLIP_FILTER_LIST = DATA_ROOT / "data/clip_image_embeddings_doc_normalized_filtered-by-clip-0.26.txt"
MONTAGE_OUT_DIR = NOTEBOOK_DIR  # where to save bv_things_correlation_montages_*.png/pdf

# Fixed categories for montage (high / low Spearman r)
HIGH_CORRELATION_CATEGORIES = ["zebra", "giraffe", "cloud", "pasta", "shoe", "stick", "dog", "cat", "ball", "bottle"]
LOW_CORRELATION_CATEGORIES = ["mop", "soap", "elephant", "bear", "slipper", "watch", "alligator", "bird", "table", "chair"]
N_CROPS, CELL_SIZE, N_COLS_MONTAGE = 10, (80, 80), 5
BODY_PARTS = {"ankle", "arm", "ear", "eye", "face", "finger", "foot", "hair", "hand", "leg", "mouth", "nose", "toe", "tooth", "neck"}

print("Configuration loaded. Please review and update paths as needed.")

Configuration loaded. Please review and update paths as needed.


## Correlate Category Embeddings

In [3]:
print("="*60)
print("CORRELATING CATEGORY EMBEDDINGS")
print("="*60)

csv1_path = Path(CSV_FILE1)
csv2_path = Path(CSV_FILE2)

if not csv1_path.exists():
    print(f"Error: {csv1_path} not found. Please check the path.")
elif not csv2_path.exists():
    print(f"Error: {csv2_path} not found. Please check the path.")
else:
    # Load CSV files
    print(f"Loading embeddings 1 from {csv1_path}...")
    df1 = pd.read_csv(csv1_path, index_col=0)
    categories1 = df1.index.tolist()
    embeddings1 = df1.values
    print(f"  Categories: {len(categories1)}, Embedding dim: {embeddings1.shape[1]}")
    
    print(f"Loading embeddings 2 from {csv2_path}...")
    df2 = pd.read_csv(csv2_path, index_col=0)
    categories2 = df2.index.tolist()
    embeddings2 = df2.values
    print(f"  Categories: {len(categories2)}, Embedding dim: {embeddings2.shape[1]}")
    
    # Check embedding dimensions
    if embeddings1.shape[1] != embeddings2.shape[1]:
        print(f"Warning: Embedding dimensions differ: {embeddings1.shape[1]} vs {embeddings2.shape[1]}")
        min_dim = min(embeddings1.shape[1], embeddings2.shape[1])
        embeddings1 = embeddings1[:, :min_dim]
        embeddings2 = embeddings2[:, :min_dim]
        print(f"  Using first {min_dim} dimensions")
    
    # Find matching categories
    categories1_set = set(categories1)
    categories2_set = set(categories2)
    matching_categories = sorted(categories1_set & categories2_set)
    
    # Filter by category include file if provided
    if CATEGORY_INCLUDE_FILE is not None:
        include_file_path = Path(CATEGORY_INCLUDE_FILE)
        if include_file_path.exists():
            with open(include_file_path, 'r') as f:
                included_categories = set(line.strip() for line in f if line.strip())
            matching_categories = [cat for cat in matching_categories if cat in included_categories]
            print(f"After filtering by include file: {len(matching_categories)} categories")
        else:
            print(f"Warning: Category include file not found: {include_file_path}")
            print("Proceeding with all matching categories.")
    
    print(f"\nMatching categories: {len(matching_categories)}")
    
    if len(matching_categories) == 0:
        print("Error: No matching categories found!")
    else:
        # Create mapping from category to index
        cat_to_idx1 = {cat: idx for idx, cat in enumerate(categories1)}
        cat_to_idx2 = {cat: idx for idx, cat in enumerate(categories2)}
        
        # Compute correlations for each matching category
        per_category_results = []
        all_pearson_rs = []
        all_spearman_rs = []
        all_kendall_rs = []
        all_cosine_sims = []
        
        for cat in matching_categories:
            idx1 = cat_to_idx1[cat]
            idx2 = cat_to_idx2[cat]
            
            vec1 = embeddings1[idx1]
            vec2 = embeddings2[idx2]
            
            # Remove NaN/Inf
            mask = np.isfinite(vec1) & np.isfinite(vec2)
            vec1_clean = vec1[mask]
            vec2_clean = vec2[mask]
            
            if len(vec1_clean) >= 3:
                pearson_r, pearson_p = pearsonr(vec1_clean, vec2_clean)
                spearman_r, spearman_p = spearmanr(vec1_clean, vec2_clean)
                kendall_r, kendall_p = kendalltau(vec1_clean, vec2_clean)
            else:
                pearson_r, pearson_p = np.nan, np.nan
                spearman_r, spearman_p = np.nan, np.nan
                kendall_r, kendall_p = np.nan, np.nan
            
            # Cosine similarity
            if len(vec1_clean) > 0:
                vec1_2d = vec1_clean.reshape(1, -1)
                vec2_2d = vec2_clean.reshape(1, -1)
                cosine_sim = cosine_similarity(vec1_2d, vec2_2d)[0, 0]
            else:
                cosine_sim = np.nan
            
            per_category_results.append({
                'category': cat,
                'pearson_r': pearson_r,
                'spearman_r': spearman_r,
                'kendall_r': kendall_r,
                'cosine_similarity': cosine_sim
            })
            
            if not np.isnan(pearson_r):
                all_pearson_rs.append(pearson_r)
            if not np.isnan(spearman_r):
                all_spearman_rs.append(spearman_r)
            if not np.isnan(kendall_r):
                all_kendall_rs.append(kendall_r)
            if not np.isnan(cosine_sim):
                all_cosine_sims.append(cosine_sim)
        
        # Summary statistics
        print("\n" + "="*60)
        print("SUMMARY STATISTICS")
        print("="*60)
        print(f"Categories analyzed: {len(matching_categories)}")
        print(f"\nPearson Correlation:")
        print(f"  Mean:   {np.nanmean(all_pearson_rs):.6f}")
        print(f"  Std:    {np.nanstd(all_pearson_rs):.6f}")
        print(f"  Median: {np.nanmedian(all_pearson_rs):.6f}")
        print(f"  Min:    {np.nanmin(all_pearson_rs):.6f}")
        print(f"  Max:    {np.nanmax(all_pearson_rs):.6f}")
        print(f"\nSpearman Correlation:")
        print(f"  Mean:   {np.nanmean(all_spearman_rs):.6f}")
        print(f"  Std:    {np.nanstd(all_spearman_rs):.6f}")
        print(f"  Median: {np.nanmedian(all_spearman_rs):.6f}")
        print(f"\nKendall Correlation:")
        print(f"  Mean:   {np.nanmean(all_kendall_rs):.6f}")
        print(f"  Std:    {np.nanstd(all_kendall_rs):.6f}")
        print(f"  Median: {np.nanmedian(all_kendall_rs):.6f}")
        print(f"\nCosine Similarity:")
        print(f"  Mean:   {np.nanmean(all_cosine_sims):.6f}")
        print(f"  Std:    {np.nanstd(all_cosine_sims):.6f}")
        print(f"  Median: {np.nanmedian(all_cosine_sims):.6f}")
        
        # Top and bottom categories
        sorted_results = sorted(per_category_results, 
                              key=lambda x: x['pearson_r'] if not np.isnan(x['pearson_r']) else -np.inf, 
                              reverse=True)
        
        print(f"\n\nTop 10 categories by Pearson correlation:")
        for i, result in enumerate(sorted_results[:10], 1):
            print(f"  {i:2d}. {result['category']:<30} r={result['pearson_r']:.6f}, cos={result['cosine_similarity']:.6f}")
        
        print(f"\nBottom 10 categories by Pearson correlation:")
        for i, result in enumerate(sorted_results[-10:], len(sorted_results)-9):
            print(f"  {i:2d}. {result['category']:<30} r={result['pearson_r']:.6f}, cos={result['cosine_similarity']:.6f}")

        # Save results to output directory
        output_dir = Path(OUTPUT_DIR)
        output_dir.mkdir(parents=True, exist_ok=True)

        # Create results text file
        results_file = output_dir / OUTPUT_FILENAME
        with open(results_file, 'w') as f:
            f.write("="*60 + "\n")
            f.write("CATEGORY EMBEDDING CORRELATION RESULTS\n")
            f.write("="*60 + "\n\n")
            f.write(f"CSV File 1: {csv1_path}\n")
            f.write(f"CSV File 2: {csv2_path}\n\n")
            f.write(f"Categories 1: {len(categories1)}, Embedding dim: {embeddings1.shape[1]}\n")
            f.write(f"Categories 2: {len(categories2)}, Embedding dim: {embeddings2.shape[1]}\n")
            if CATEGORY_INCLUDE_FILE is not None:
                f.write(f"Category include file: {CATEGORY_INCLUDE_FILE}\n")
            f.write(f"Matching categories: {len(matching_categories)}\n\n")
            
            f.write("="*60 + "\n")
            f.write("SUMMARY STATISTICS\n")
            f.write("="*60 + "\n")
            f.write(f"Categories analyzed: {len(matching_categories)}\n\n")
            
            f.write("Pearson Correlation:\n")
            f.write(f"  Mean:   {np.nanmean(all_pearson_rs):.6f}\n")
            f.write(f"  Std:    {np.nanstd(all_pearson_rs):.6f}\n")
            f.write(f"  Median: {np.nanmedian(all_pearson_rs):.6f}\n")
            f.write(f"  Min:    {np.nanmin(all_pearson_rs):.6f}\n")
            f.write(f"  Max:    {np.nanmax(all_pearson_rs):.6f}\n\n")
            
            f.write("Spearman Correlation:\n")
            f.write(f"  Mean:   {np.nanmean(all_spearman_rs):.6f}\n")
            f.write(f"  Std:    {np.nanstd(all_spearman_rs):.6f}\n")
            f.write(f"  Median: {np.nanmedian(all_spearman_rs):.6f}\n")
            f.write(f"  Min:    {np.nanmin(all_spearman_rs):.6f}\n")
            f.write(f"  Max:    {np.nanmax(all_spearman_rs):.6f}\n\n")
            
            f.write("Kendall Correlation:\n")
            f.write(f"  Mean:   {np.nanmean(all_kendall_rs):.6f}\n")
            f.write(f"  Std:    {np.nanstd(all_kendall_rs):.6f}\n")
            f.write(f"  Median: {np.nanmedian(all_kendall_rs):.6f}\n")
            f.write(f"  Min:    {np.nanmin(all_kendall_rs):.6f}\n")
            f.write(f"  Max:    {np.nanmax(all_kendall_rs):.6f}\n\n")
            
            f.write("Cosine Similarity:\n")
            f.write(f"  Mean:   {np.nanmean(all_cosine_sims):.6f}\n")
            f.write(f"  Std:    {np.nanstd(all_cosine_sims):.6f}\n")
            f.write(f"  Median: {np.nanmedian(all_cosine_sims):.6f}\n")
            f.write(f"  Min:    {np.nanmin(all_cosine_sims):.6f}\n")
            f.write(f"  Max:    {np.nanmax(all_cosine_sims):.6f}\n\n")
            
            f.write("="*60 + "\n")
            f.write("TOP 10 CATEGORIES BY PEARSON CORRELATION\n")
            f.write("="*60 + "\n")
            for i, result in enumerate(sorted_results[:10], 1):
                f.write(f"  {i:2d}. {result['category']:<30} pearson_r={result['pearson_r']:.6f}, spearman_r={result['spearman_r']:.6f}, kendall_r={result['kendall_r']:.6f}, cosine={result['cosine_similarity']:.6f}\n")
            
            f.write("\n" + "="*60 + "\n")
            f.write("BOTTOM 10 CATEGORIES BY PEARSON CORRELATION\n")
            f.write("="*60 + "\n")
            for i, result in enumerate(sorted_results[-10:], len(sorted_results)-9):
                f.write(f"  {i:2d}. {result['category']:<30} pearson_r={result['pearson_r']:.6f}, spearman_r={result['spearman_r']:.6f}, kendall_r={result['kendall_r']:.6f}, cosine={result['cosine_similarity']:.6f}\n")
        
        print(f"\nResults saved to: {results_file}")


CORRELATING CATEGORY EMBEDDINGS
Loading embeddings 1 from /home/j7yang/babyview-projects/vss2026/object-detection/analysis/vss-2026/bv_things_comp_12252025/bv_dinov3_filtered_zscored_hierarchical_163cats/normalized_filtered_embeddings_alphabetical.csv...
  Categories: 163, Embedding dim: 768
Loading embeddings 2 from /home/j7yang/babyview-projects/vss2026/object-detection/analysis/vss-2026/bv_things_comp_12252025/things_dinov3_filtered_zscored_hierarchical_163cats/normalized_filtered_embeddings_alphabetical.csv...
  Categories: 163, Embedding dim: 768
After filtering by include file: 163 categories

Matching categories: 163

SUMMARY STATISTICS
Categories analyzed: 163

Pearson Correlation:
  Mean:   0.367984
  Std:    0.166945
  Median: 0.366334
  Min:    -0.013829
  Max:    0.729533

Spearman Correlation:
  Mean:   0.352944
  Std:    0.161349
  Median: 0.351250

Kendall Correlation:
  Mean:   0.243660
  Std:    0.115991
  Median: 0.236500

Cosine Similarity:
  Mean:   0.367907
  Std: 

## CLIP vs DINOv3 category-wise correlation

For each category we have a BV–THINGS correlation from CLIP and one from DINOv3. This section computes how correlated those two vectors are across the 163 categories: do categories with high CLIP agreement tend to have high DINOv3 agreement (and vice versa)? Uses the same embedding CSVs as the correlation reports.

In [4]:
# CLIP vs DINOv3: correlation of category-wise BV–THINGS correlations (from correlation_clip_vs_dinov3.py)
from scipy import stats

def load_embeddings_csv_for_pearson(csv_path):
    """Load embedding CSV: index = category, columns = dim_*."""
    df = pd.read_csv(csv_path, index_col=0)
    if df.index.name is None or str(df.index.name).startswith("Unnamed"):
        if "Unnamed: 0" in df.columns:
            df = df.set_index("Unnamed: 0")
    dim_cols = [c for c in df.columns if re.match(r"dim_\d+", str(c))]
    if not dim_cols:
        dim_cols = [c for c in df.columns if c != "category"]
    X = df[dim_cols].values.astype(np.float64)
    categories = [str(i).strip().lower() for i in df.index]
    return categories, X

def compute_pearson_per_category(bv_csv, things_csv, category_include_file=None):
    """Return dict category -> pearson_r for BV vs THINGS embeddings."""
    cats_bv, X_bv = load_embeddings_csv_for_pearson(bv_csv)
    cats_th, X_th = load_embeddings_csv_for_pearson(things_csv)
    matching = [c for c in cats_bv if c in set(cats_th)]
    if category_include_file and Path(category_include_file).exists():
        with open(category_include_file) as f:
            included = set(line.strip().lower() for line in f if line.strip())
        matching = [c for c in matching if c in included]
    cat_to_idx_bv = {c: i for i, c in enumerate(cats_bv)}
    cat_to_idx_th = {c: i for i, c in enumerate(cats_th)}
    out = {}
    for cat in matching:
        v1 = X_bv[cat_to_idx_bv[cat]]
        v2 = X_th[cat_to_idx_th[cat]]
        mask = np.isfinite(v1) & np.isfinite(v2)
        v1c, v2c = v1[mask], v2[mask]
        r, _ = stats.pearsonr(v1c, v2c) if len(v1c) >= 3 else (np.nan, np.nan)
        out[cat] = r
    return out

clip_r = compute_pearson_per_category(BV_CLIP_CSV, THINGS_CLIP_CSV, CATEGORY_INCLUDE_FILE)
dinov3_r = compute_pearson_per_category(BV_DINOV3_CSV, THINGS_DINOV3_CSV, CATEGORY_INCLUDE_FILE)
common = sorted(set(clip_r) & set(dinov3_r))
clip_vals = np.array([clip_r[c] for c in common])
dinov3_vals = np.array([dinov3_r[c] for c in common])
valid = np.isfinite(clip_vals) & np.isfinite(dinov3_vals)
n_valid = int(np.sum(valid))
clip_clean = clip_vals[valid]
dinov3_clean = dinov3_vals[valid]

pearson_r, pearson_p = stats.pearsonr(clip_clean, dinov3_clean)
spearman_r, spearman_p = stats.spearmanr(clip_clean, dinov3_clean)

print("=" * 60)
print("CLIP vs DINOv3 category-wise BV–THINGS correlations")
print("=" * 60)
print(f"Categories (shared, valid): {n_valid} of {len(common)}")
print()
print("Correlation across categories (CLIP r vs DINOv3 r):")
print(f"  Pearson  r = {pearson_r:.4f}, p = {pearson_p:.2e}")
print(f"  Spearman r = {spearman_r:.4f}, p = {spearman_p:.2e}")
print()
print("Interpretation: categories that have higher BV–THINGS agreement")
print("in CLIP tend to have higher agreement in DINOv3 (and vice versa).")
print("=" * 60)

OUTPUT_DIR.mkdir(parents=True, exist_ok=True)
with open(CLIP_VS_DINOV3_CSV, "w") as f:
    f.write("category,clip_pearson_r,dinov3_pearson_r\n")
    for c in common:
        f.write(f"{c},{clip_r[c]:.6f},{dinov3_r[c]:.6f}\n")
print(f"Wrote aligned table: {CLIP_VS_DINOV3_CSV}")

Found 163 categories in inclusion file
Precision file: 320 rows, annotators: ['Dora', 'Mira']
Rows after filtering to included categories: 148

Categories: 113, global average precision: 0.597149, std: 0.261254
Saved to /home/j7yang/babyview-projects/vss2026/object-detection/data/average_precision_for_included_categories.csv


## BV–THINGS correlation montages and histograms

For each embedding model (CLIP and DINOv3), builds a figure: left = montages of high/low Spearman-r categories (BV | THINGS), right = histogram of correlation distribution. Requires `CROPPED_DIR` and `THINGS_IMAGES_DIR` to exist for image montages; otherwise histograms only.

In [5]:
# --- Montage helpers (from plot_bv_things_correlation_montages.py) ---
def load_embeddings_csv(csv_path):
    df = pd.read_csv(csv_path, index_col=0)
    if df.index.name is None or str(df.index.name).startswith("Unnamed"):
        if "Unnamed: 0" in df.columns:
            df = df.set_index("Unnamed: 0")
    dim_cols = [c for c in df.columns if re.match(r"dim_\d+", str(c))]
    if not dim_cols:
        dim_cols = [c for c in df.columns if c != "category"]
    X = df[dim_cols].values.astype(np.float64)
    categories = [str(i).strip() for i in df.index]
    return categories, X

def compute_spearman_per_category(bv_csv, things_csv, category_include_file=None):
    cats_bv, X_bv = load_embeddings_csv(bv_csv)
    cats_th, X_th = load_embeddings_csv(things_csv)
    cats_bv = [c.strip().lower() for c in cats_bv]
    cats_th = [c.strip().lower() for c in cats_th]
    matching = [c for c in cats_bv if c in set(cats_th)]
    if category_include_file and Path(category_include_file).exists():
        with open(category_include_file) as f:
            included = set(line.strip().lower() for line in f if line.strip())
        matching = [c for c in matching if c in included]
    cat_to_idx_bv = {c: i for i, c in enumerate(cats_bv)}
    cat_to_idx_th = {c: i for i, c in enumerate(cats_th)}
    results, all_rs = [], []
    for cat in matching:
        v1 = X_bv[cat_to_idx_bv[cat]]
        v2 = X_th[cat_to_idx_th[cat]]
        mask = np.isfinite(v1) & np.isfinite(v2)
        v1c, v2c = v1[mask], v2[mask]
        r, _ = spearmanr(v1c, v2c) if len(v1c) >= 3 else (np.nan, np.nan)
        results.append({"category": cat, "spearman_r": r})
        if not np.isnan(r):
            all_rs.append(r)
    return results, all_rs

def get_category_crop_dir(root_dir, cat_name):
    root_dir = Path(root_dir)
    cat_lower = cat_name.strip().lower()
    direct = root_dir / cat_name
    if direct.exists() and direct.is_dir():
        return direct
    if root_dir.exists():
        for p in root_dir.iterdir():
            if p.is_dir() and p.name.lower() == cat_lower:
                return p
    return root_dir / cat_name

def load_clip_filtered_stems_for_categories(filter_list_path, categories_set):
    categories_set = {c.strip().lower() for c in categories_set}
    allowed = set()
    if not Path(filter_list_path).exists():
        return allowed
    with open(filter_list_path, "r") as f:
        for line in f:
            line = line.strip()
            if not line:
                continue
            p = Path(line)
            cat = p.parent.name.strip().lower()
            if cat not in categories_set:
                continue
            stem = p.stem if p.suffix.lower() == ".npy" else p.name
            allowed.add((cat, stem))
    return allowed

def get_bv_crop_paths(cropped_dir, cat_name, n=10, allowed_stems=None):
    cat_dir = get_category_crop_dir(cropped_dir, cat_name)
    if not cat_dir.exists():
        return []
    paths = []
    for ext in ("*.jpg", "*.jpeg", "*.png"):
        paths.extend(cat_dir.glob(ext))
    cat_lower = cat_name.strip().lower()
    if allowed_stems is not None:
        paths = [p for p in paths if (cat_lower, p.stem) in allowed_stems]
    return sorted(paths, key=lambda p: p.name)[:n]

def get_things_crop_paths(things_images_dir, cat_name, n=10):
    if things_images_dir is None or not Path(things_images_dir).exists():
        return []
    cat_dir = get_category_crop_dir(things_images_dir, cat_name)
    if not cat_dir.exists():
        return []
    paths = []
    for ext in ("*.jpg", "*.jpeg", "*.png"):
        paths.extend(cat_dir.glob(ext))
    return sorted(paths, key=lambda p: p.name)[:n]

def make_montage(paths, cell_size, n_cols):
    if not paths:
        return None
    n = len(paths)
    n_rows = (n + n_cols - 1) // n_cols
    out = Image.new("RGB", (n_cols * cell_size[0], n_rows * cell_size[1]), (240, 240, 240))
    for idx, p in enumerate(paths):
        try:
            img = Image.open(p).convert("RGB")
            if img.size != cell_size:
                img = img.resize(cell_size, Image.Resampling.LANCZOS)
            row, col = idx // n_cols, idx % n_cols
            out.paste(img, (col * cell_size[0], row * cell_size[1]))
        except Exception:
            pass
    return out

def build_montage_figure(model_name, results_sorted, all_spearman_rs, cropped_dir, things_images_dir,
    n_high=5, n_low=5, n_crops=N_CROPS, cell_size=CELL_SIZE, n_cols=N_COLS_MONTAGE,
    allowed_stems=None, fixed_high_categories=None, fixed_low_categories=None):
    cat_to_result = {r["category"].strip().lower(): r for r in results_sorted}
    if fixed_high_categories is not None and fixed_low_categories is not None:
        rows = []
        for cat in fixed_high_categories + fixed_low_categories:
            c = cat.strip().lower()
            rows.append(cat_to_result.get(c, {"category": cat, "spearman_r": np.nan}))
    else:
        valid = [r for r in results_sorted if not np.isnan(r["spearman_r"]) and r["category"].strip().lower() not in BODY_PARTS]
        rows = valid[-n_high:][::-1] + valid[:n_low]
    n_rows = len(rows)
    fig = plt.figure(figsize=(12, 2.2 * n_rows))
    gs = GridSpec(n_rows, 3, figure=fig, width_ratios=[1.2, 1.2, 0.7], hspace=0.28, wspace=0.05)
    for i, row in enumerate(rows):
        cat, r_val = row["category"], row["spearman_r"]
        bv_paths = get_bv_crop_paths(cropped_dir, cat, n_crops, allowed_stems=allowed_stems)
        th_paths = get_things_crop_paths(things_images_dir, cat, n_crops)
        ax_bv = fig.add_subplot(gs[i, 0])
        ax_th = fig.add_subplot(gs[i, 1])
        if bv_paths:
            mont_bv = make_montage(bv_paths, cell_size, n_cols)
            if mont_bv is not None:
                ax_bv.imshow(np.array(mont_bv))
        ax_bv.axis("off")
        ax_bv.set_title(f"{cat}, r = {r_val:.2f}", fontsize=9)
        if th_paths:
            mont_th = make_montage(th_paths, cell_size, n_cols)
            if mont_th is not None:
                ax_th.imshow(np.array(mont_th))
        ax_th.axis("off")
    fig.text(0.22, 0.96, "BV", fontsize=10, ha="center")
    fig.text(0.50, 0.96, "THINGS", fontsize=10, ha="center")
    ax_hist = fig.add_subplot(gs[:, 2])
    rs = [x for x in all_spearman_rs if not np.isnan(x)]
    if rs:
        ax_hist.hist(rs, bins=20, orientation="horizontal", color="steelblue", alpha=0.8, edgecolor="white")
        ax_hist.axhline(y=np.median(rs), color="gray", linestyle="--", linewidth=1, label=f"Median = {np.median(rs):.2f}")
        ax_hist.set_ylabel("Correlation (Spearman r)")
        ax_hist.set_xlabel("Count")
        ax_hist.legend(loc="upper left", fontsize=8)
    ax_hist.set_title("Correlation distribution", fontsize=10)
    fig.suptitle(f"BV–THINGS category correlation variability ({model_name})\n{n_high} high + {n_low} low Spearman r", fontsize=12, y=0.98)
    return fig

In [6]:
# Generate and save montage figures for CLIP and DINOv3
Path(MONTAGE_OUT_DIR).mkdir(parents=True, exist_ok=True)
include_file = str(CATEGORY_INCLUDE_FILE) if CATEGORY_INCLUDE_FILE else None
use_clip_filter = CLIP_FILTER_LIST and Path(CLIP_FILTER_LIST).exists()

for model_name, bv_csv, th_csv in [
    ("CLIP", BV_CLIP_CSV, THINGS_CLIP_CSV),
    ("DINOv3", BV_DINOV3_CSV, THINGS_DINOV3_CSV),
]:
    if not Path(bv_csv).exists() or not Path(th_csv).exists():
        print(f"Skipping {model_name}: missing CSV")
        continue
    results, all_rs = compute_spearman_per_category(bv_csv, th_csv, include_file)
    results_sorted = sorted(results, key=lambda x: x["spearman_r"] if not np.isnan(x["spearman_r"]) else -np.inf)
    categories_needed = {c.strip().lower() for c in HIGH_CORRELATION_CATEGORIES + LOW_CORRELATION_CATEGORIES}
    allowed_stems = load_clip_filtered_stems_for_categories(CLIP_FILTER_LIST, categories_needed) if use_clip_filter else None
    if CROPPED_DIR and not Path(CROPPED_DIR).exists():
        print(f"Warning: CROPPED_DIR not found: {CROPPED_DIR}. Montages will be empty.")
    if THINGS_IMAGES_DIR and (not Path(THINGS_IMAGES_DIR).exists()):
        print(f"Warning: THINGS_IMAGES_DIR not found: {THINGS_IMAGES_DIR}.")
    fig = build_montage_figure(
        model_name, results_sorted, all_rs, CROPPED_DIR, THINGS_IMAGES_DIR,
        n_high=len(HIGH_CORRELATION_CATEGORIES), n_low=len(LOW_CORRELATION_CATEGORIES),
        n_crops=N_CROPS, cell_size=CELL_SIZE, n_cols=N_COLS_MONTAGE,
        allowed_stems=allowed_stems,
        fixed_high_categories=HIGH_CORRELATION_CATEGORIES,
        fixed_low_categories=LOW_CORRELATION_CATEGORIES,
    )
    base_name = f"bv_things_correlation_montages_{model_name.lower()}"
    for ext, opts in [("png", {"dpi": 150}), ("pdf", {})]:
        out_path = Path(MONTAGE_OUT_DIR) / f"{base_name}.{ext}"
        fig.savefig(out_path, bbox_inches="tight", **opts)
        print(f"Saved {out_path}")
    plt.close(fig)
print("Done.")

Saved /home/j7yang/babyview-projects/vss2026/object-detection/analysis/vss-2026/bv_things_correlation_montages_clip.png
Saved /home/j7yang/babyview-projects/vss2026/object-detection/analysis/vss-2026/bv_things_correlation_montages_clip.pdf
Saved /home/j7yang/babyview-projects/vss2026/object-detection/analysis/vss-2026/bv_things_correlation_montages_dinov3.png
Saved /home/j7yang/babyview-projects/vss2026/object-detection/analysis/vss-2026/bv_things_correlation_montages_dinov3.pdf
Done.
